include/alpaka/core/Assert.hpp

```diff
@@ -71,41 +62,26 @@ namespace alpaka
 
         namespace detail
         {
-            template<typename TLhs, typename TRhs, typename TSfinae = void>
-            struct AssertGreaterThan;
             template<typename TLhs, typename TRhs>
-            struct AssertGreaterThan<
-                TLhs,
-                TRhs,
-                std::enable_if_t<!std::is_unsigned<TRhs>::value || (TLhs::value != 0u)>>
+            struct AssertGreaterThan
             {
-                ALPAKA_NO_HOST_ACC_WARNING
-                ALPAKA_FN_HOST_ACC static auto assertGreaterThan(TRhs const& lhs) -> void
+                ALPAKA_NO_HOST_ACC_WARNING ALPAKA_FN_HOST_ACC static auto assertGreaterThan(TRhs const& rhs)
                 {
-                    alpaka::ignore_unused(lhs);
-                    ALPAKA_ASSERT_OFFLOAD(TLhs::value > lhs);
-                }
-            };
-            template<typename TLhs, typename TRhs>
-            struct AssertGreaterThan<
-                TLhs,
-                TRhs,
-                std::enable_if_t<std::is_unsigned<TRhs>::value && (TLhs::value == 0u)>>
-            {
-                ALPAKA_NO_HOST_ACC_WARNING
-                ALPAKA_FN_HOST_ACC static auto assertGreaterThan(TRhs const& lhs) -> void
-                {
-                    alpaka::ignore_unused(lhs);
-                    // Nothing to do for unsigned types camparing to zero.
+                    alpaka::ignore_unused(rhs);
+                    if constexpr(std::is_signed_v<TRhs> || (TLhs::value != 0u))
```

Comment:
I wonder if `TLhs` here should also be `TRhs`.

Reply:
Nope, `TLhs` is fine here. The check is only used inside `Vec.hpp` and ensures that the user doesn't request a vector element outside of the vector dimensionality.

Reply:
I think the logic is to avoid comparison `0 > unsigned_type_value`. So the condition is the opposite of this case

Reply:
Well, the struct is called `AssertGreaterThan`. But since a `Vec<0>` class defines `Lhs = 0` the check is not applicable because it would fail for every possible vector index including `0`. Looks a bit like an oversight to me, we should probably ensure that users don't call `Vec<0>[1]`.